feat: support month and year interval for date_bin on constant data

NGA-TRAN opened the pull request:
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/5689


# Rationale for this change

In IOx, we see users use `data_bin` with month interval to group data by month for their monthly report. We think it is a very useful feature to support.

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

# What changes are included in this PR?

Since each month does not include the same number of nanoseconds and for date_bin, it also depends on the `origin` input. This PR mainly add a function to handle month interval for date bin and some related refactoring work

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

# Are these changes tested?

Yes

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

# Are there any user-facing changes?

Yes, date_bin with month and year interval is now working on constant input

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please add the `api change` label.
-->

NGA-TRAN commented on datafusion/core/tests/sqllogictests/test_files/timestamps.slt:
```diff
@@ -500,6 +500,118 @@ FROM (
     (TIMESTAMP '2021-06-10 17:19:10Z', TIMESTAMP '2001-01-01T00:00:00Z', 0.3)
   ) as t (time, origin, val)
 
+# MONTH/YEAR INTERVAL ON ARRAY DATA STILL NOT PROVIDE CORRECT ANSWERS YET
```
Test on array data (equivalent to testing on table data). These tests do not yet return the correct answers. See below for my comment on how to fix this in my next PR

NGA-TRAN commented on datafusion/core/tests/sqllogictests/test_files/timestamps.slt:
```diff
@@ -500,6 +500,118 @@ FROM (
+2021-04-03T00:00:00
+2022-03-29T00:00:00
+
+# month interval on constant
```
From here down, all the tests did not work before (error: not supported) but now working

NGA-TRAN commented on datafusion/physical-expr/src/datetime_expressions.rs:
```diff
@@ -333,21 +333,87 @@ pub fn date_trunc(args: &[ColumnarValue]) -> Result<ColumnarValue> {
-fn date_bin_single(stride: i64, source: i64, origin: i64) -> i64 {
+trait DateBinInterval {
+    fn date_bin_interval(&self, stride: i64, source: i64, origin: i64) -> i64;
+}
```
I create this trait for it to work with an available closure because I now need 2 different ways to handle date_bin  intervals nanoseconds and months

NGA-TRAN replied:
This trait is no longer needed from the comment below

NGA-TRAN commented on datafusion/physical-expr/src/datetime_expressions.rs:
```diff
@@ -429,7 +515,18 @@ fn date_bin_impl(
+                y.date_bin_interval(stride, x, origin)
+            }
+        })
+    };
```
This is the closure

NGA-TRAN commented on datafusion/core/tests/sqllogictests/test_files/timestamps.slt:
```diff
@@ -421,7 +421,7 @@ SELECT DATE_BIN(INTERVAL '5 microseconds', TIMESTAMP '2022-08-03 14:38:50.000006
 
 # Does not support months for Month-Day-Nano interval
-statement error This feature is not implemented: DATE_BIN stride does not support month intervals
+statement error DataFusion error: This feature is not implemented: DATE_BIN stride does not support combination of month, day and nanosecond intervals
```
new error message I have added in this PR

NGA-TRAN commented on datafusion/physical-expr/src/datetime_expressions.rs:
```diff
@@ -366,6 +432,17 @@ pub fn date_bin(args: &[ColumnarValue]) -> Result<ColumnarValue> {
+//  1. IntervalDayTime: this means that the stride is in days, hours, minutes, seconds and milliseconds
+//     We will assume month interval won't be converted into this type
+//     TODO (my next PR): for array data, the stride was converted into ScalarValue::IntervalDayTime somwhere
+//             for month interval. I need to find that and make it ScalarValue::IntervalMonthDayNano instead
```
This is how to support month interval on array data. I think the change will be minimal but I need to figure out where

NGA-TRAN commented:
@alamb

NGA-TRAN commented on datafusion/physical-expr/src/datetime_expressions.rs:
```diff
@@ -333,21 +333,87 @@ pub fn date_trunc(args: &[ColumnarValue]) -> Result<ColumnarValue> {
+        + source_date.month() as i32
+        - origin_date.month() as i32;
+
+    // distance from origin to bin
+    let month_delta = month_diff - (month_diff % stride_months as i32);
+
+    let month_delta = if month_diff < 0 && stride_months > 1 {
+        // The origin is later than the source timestamp, round down to the previous bin
+        month_delta - stride_months as i32
+    } else {
+        month_delta
+    };
```
I will refactor putting these line into a function to get used here and in the above function

NGA-TRAN replied:
Done

alamb commented:
> The month interval for array input was converted into ScalarValue::IntervalDayTime (days) somewhere. My next PR is to find that place and convert month interval into ScalarValue::IntervalMonthDayNano instead. I think if I do that, they will automatically work with this code


I suspect the code that is do this is the coercion to the type signature defined here:

https://github.com/apache/arrow-datafusion/blob/1646bf6aae1e568e38c5a43e9a44c34e26222341/datafusion/expr/src/function.rs#L474-L494

I think you could move the `Interval(MonthDayNano)` signatures above the `Interval(DayTime)` ones

alamb commented on datafusion/physical-expr/src/datetime_expressions.rs:
```diff
@@ -366,6 +432,17 @@ pub fn date_bin(args: &[ColumnarValue]) -> Result<ColumnarValue> {
     }
 }
 
+enum Interval {
```
Since we have this enum anyways, I wonder if there is any need for the trait?

For example, what if the code looked like

```rust
impl Inteval {
  // return time in nanoseconds that the source timestamp falls into based on the stride and origin
  fn bin(source: i64, origin: i64) -> i64 {
    match self {
      Self::Nanoseconds(v) => date_bin_nanos_interval(v, source, origin),
      Self::Months(v) => date_bin_months_interval(v, source, origin)
    }
  }
```

And then the  closure looks like
```
    let f = |x: Option<i64>| x.map(move |x| stride.bin( x, origin));
```

🤔 

I don't think you would need a trait at all

NGA-TRAN replied:
It works. Thanks Andrew

alamb reviewed (commented):
Looks like a good direction to me @NGA-TRAN   -- thank you.

I left some comments that I think will help simplify the code

FYI @berkaysynnada  @waitingkuo @mustafasrepo  and @metesynnada  who have been working in the interval area recently

NGA-TRAN commented on datafusion/core/tests/sqllogictests/test_files/timestamps.slt:
```diff
@@ -500,6 +500,118 @@ FROM (
+
+# month interval on constant
+query P
+SELECT DATE_BIN(INTERVAL '1 month', TIMESTAMP '2022-01-01 00:00:00Z', TIMESTAMP '1970-01-01T00:00:00Z');
```
Hah, by removing these, I have found that without keyword `INTERVAL`, the month is converted into days even on constant time. That is the problem I will work on next. Array data works correctly if there is `INTERVAL` keyword

NGA-TRAN commented on datafusion/physical-expr/src/datetime_expressions.rs:
```diff
@@ -333,21 +333,87 @@ pub fn date_trunc(args: &[ColumnarValue]) -> Result<ColumnarValue> {
+}
+
+// return time in nanoseconds that the source timestamp falls into based on the stride and origin
+fn date_bin_nanos_interval(stride_nanos: i64, source: i64, origin: i64) -> i64 {
```
This is also no longer needed after removing the trait

NGA-TRAN commented on datafusion/core/tests/sqllogictests/test_files/timestamps.slt:
```diff
@@ -500,6 +500,268 @@ FROM (
+2021-12-29T00:00:00
+
+
+#################
```
From here down having `INTERVAL` and return correct results

NGA-TRAN reviewed (commented):
Thanks @alamb . I have address your comments. I also fixed a bug I just found. It was about origin the is not first moment of the month

NGA-TRAN commented:
@alamb

> I think you could move the Interval(MonthDayNano) signatures above the Interval(DayTime) ones

It works like a charm. This PR now has dateline works for everything (if I do not miss other cases).
To avoid many duplicate tests with and without INTERVAL keyword, I added INTERVAL into tests without it before and remove it for the ones with it. These make the tests more concise and I know all of them work correctly

alamb commented on datafusion/core/tests/sqllogictests/test_files/timestamps.slt:
```diff
@@ -500,6 +500,267 @@ FROM (
+2022-02-01T00:00:00
+2022-03-01T00:00:00
+
+# month interval with specified start time
```
in this case the specified start time is the same as the default '1970-01-01T00:00:00Z'. Do you think it is adding useful coverge?

NGA-TRAN replied:
Somehow I missed many comments in previous run. I do not know the DF code very well. If you say they go thru the same code path, I will remove this

NGA-TRAN replied:
Done

alamb commented on datafusion/core/tests/sqllogictests/test_files/timestamps.slt:
```diff
@@ -500,6 +500,267 @@ FROM (
+2022-01-01T00:00:00
+
+
+# Tests without INTERVAL keyword
```
given this is a test for date_bin I am not sure why it is useful to test both with and without `INTERVAL`. Test maintenance has been on my mind lately which is why I am asking about the need for certain tests

NGA-TRAN replied:
I add tests for both with and without INTERVAL because they did not work the same before.

NGA-TRAN replied:
Done

alamb commented on datafusion/core/tests/sqllogictests/test_files/timestamps.slt:
```diff
@@ -500,6 +500,267 @@ FROM (
+
+# 2-month interval in date_bin with default start time
+query P
+select date_bin('2 month', column1)
```
👍

alamb commented on datafusion/core/tests/sqllogictests/test_files/timestamps.slt:
```diff
@@ -500,6 +500,267 @@ FROM (
+
+# month interval with specified start time
+query P
+select date_bin('1 month', column1, '1970-01-01T00:00:00Z')
```
as above it is not clear why this particular variation is needed

NGA-TRAN replied:
I will remove this shortly, too

NGA-TRAN replied:
Done

alamb commented on datafusion/core/tests/sqllogictests/test_files/timestamps.slt:
```diff
@@ -500,6 +500,267 @@ FROM (
+2022-03-01T00:00:00
+
+# month interval with start date is end of the month plus some minutes
+# Note the datetime '2022-03-31 00:00:00'. Its bin is NOT '2022-03-31 00:15:00' which is after its time
```
👍

alamb commented on datafusion/core/tests/sqllogictests/test_files/timestamps.slt:
```diff
@@ -500,6 +500,267 @@ FROM (
+  (timestamp '2022-01-02 00:00:00'), 
+  (timestamp '2022-02-02 00:00:00'),
+  (timestamp '2022-02-15 00:00:00'),
+  (timestamp '2022-03-31 00:00:00') 
```
I think this test data should perhaps also include a timestamp with a different year (`2023` perhaps) to demonstrate a second distinct bin

NGA-TRAN replied:
Added

alamb commented on datafusion/core/tests/sqllogictests/test_files/timestamps.slt:
```diff
@@ -500,6 +500,267 @@ FROM (
+----
+2021-12-31T00:15:00
+2021-12-31T00:15:00
+2021-12-31T00:15:00
```
I don't understand why the timestamp 

```
'2022-01-02 00:00:00'
``` 

would be binned into the prior year (2021)

```
2021-12-31
```

NGA-TRAN replied:
The return data is the start of the bin. The bin width is one year. The source data must be inside the bin. Since the origin is '1970-12-31T00:15:00Z', the start of the bins are '1970-12-31T00:15:00Z', '1971-12-31T00:15:00Z', ..., '2021-12-31T00:15:00Z', '2022-12-31T00:15:00Z', ...

'2022-01-02 00:00:00' is between '2021-12-31T00:15:00Z' and '2022-12-31T00:15:00Z' so it must be in the bin that starts with '2021-12-31T00:15:00Z'

alamb commented on datafusion/core/tests/sqllogictests/test_files/timestamps.slt:
```diff
@@ -500,6 +500,267 @@ FROM (
+----
+2022-01-01T00:00:00
+
+# three months interval on constant
```
the test is actually "5 months" but the comment says three

alamb commented on datafusion/core/tests/sqllogictests/test_files/timestamps.slt:
```diff
@@ -500,6 +500,267 @@ FROM (
+----
+2022-01-01T00:00:00
+
+# origin is May 51 to produce time for source Feb 28
```
I am not sure what May 51 means. Maybe this?

```suggestion
# origin is May 5 to produce time for source Feb 28
```

May51 appears below as well

NGA-TRAN replied:
Good catch. Copy and paste mistake. I have make the comments reflect the purpose of this test and the test below, too

alamb commented on datafusion/physical-expr/src/datetime_expressions.rs:
```diff
@@ -333,19 +333,73 @@ pub fn date_trunc(args: &[ColumnarValue]) -> Result<ColumnarValue> {
 
-    origin + time_delta
+    // If origin is not midnight of first date of the month, the bin_time may be larger than the source
+    // In this case, we need to move back to previous bin
```
I don't understand why we need to check for the first date of the bins

When I removed this check
```diff
diff --git a/datafusion/physical-expr/src/datetime_expressions.rs b/datafusion/physical-expr/src/datetime_expressions.rs
index 680f857168..587a12861d 100644
--- a/datafusion/physical-expr/src/datetime_expressions.rs
+++ b/datafusion/physical-expr/src/datetime_expressions.rs
@@ -377,12 +377,7 @@ fn date_bin_months_interval(stride_months: i64, source: i64, origin: i64) -> i64
 
     // If origin is not midnight of first date of the month, the bin_time may be larger than the source
     // In this case, we need to move back to previous bin
-    if (origin_date.day() != 1
-        || origin_date.hour() != 0
-        || origin_date.minute() != 0
-        || origin_date.second() != 0
-        || origin_date.nanosecond() != 0)
-        && bin_time > source_date
+    if bin_time > source_date
```

The newly added sqllogictest still passes 🤔

NGA-TRAN replied:
Agree, `if bin_time > source_date` is good enough. I have modified it

alamb commented on datafusion/physical-expr/src/datetime_expressions.rs:
```diff
@@ -429,7 +514,7 @@ fn date_bin_impl(
     };
 
-    let f = |x: Option<i64>| x.map(|x| date_bin_single(stride, x, origin));
+    let f = |x: Option<i64>| x.map(|x| stride.bin(x, origin));
```
👍  this is looking good

alamb reviewed (commented):
Thanks @NGA-TRAN  -- I reviewed the PR carefully and it is looking good. I had some questions about the tests, some of which look incorrect -- I think we should resolve the question about incorrect results prior to merging as I think having incorrect results, even if just temporarily, is not a good thing

NGA-TRAN reviewed (commented):
I have explained the test results and addressed your comments

NGA-TRAN commented:
@alamb The PR is now ready for re-reviewing with all questions answered

alamb reviewed (approved):
LGTM -- thank you @NGA-TRAN 

I took the liberty of merging this PR from main and adding a few test cases / comment clarifications in 0dc4329038

Hopefully this is ok and I plan to merge this PR once the tests pass

cc @waitingkuo @stuartcarnie

waitingkuo commented:
@alamb thank you for pining me. @NGA-TRAN  thank you for the great works. the codes looks good to me. 

i tested this in your branch, i wonder whether this corner cases make sense

```bash
❯ select date_bin('1 month', timestamp '2000-02-29T00:00:00', timestamp '2000-01-31T00:00:00');
+----------------------------------------------------------------------------------+
| datebin(Utf8("1 month"),Utf8("2000-02-29T00:00:00"),Utf8("2000-01-31T00:00:00")) |
+----------------------------------------------------------------------------------+
| 2000-02-29T00:00:00                                                              |
+----------------------------------------------------------------------------------+
1 row in set. Query took 0.003 seconds.
```

```bash
❯ select date_bin('1 month', timestamp '2000-01-31T00:00:00', timestamp '2000-02-29T00:00:00');
+----------------------------------------------------------------------------------+
| datebin(Utf8("1 month"),Utf8("2000-01-31T00:00:00"),Utf8("2000-02-29T00:00:00")) |
+----------------------------------------------------------------------------------+
| 2000-01-29T00:00:00                                                              |
+----------------------------------------------------------------------------------+
1 row in set. Query took 0.003 seconds.
```

i feel like the mechanism behind isn't consistent. i thought that the second's is `2000-01-31T00:00:00`

waitingkuo commented on datafusion/physical-expr/src/datetime_expressions.rs:
```diff
@@ -333,19 +333,67 @@ pub fn date_trunc(args: &[ColumnarValue]) -> Result<ColumnarValue> {
-    origin + time_delta
+    // If origin is not midnight of first date of the month, the bin_time may be larger than the source
+    // In this case, we need to move back to previous bin
+    if bin_time > source_date {
+        let month_delta = month_delta - stride_months;
+        bin_time = if month_delta < 0 {
+            origin_date - Months::new(month_delta.unsigned_abs() as u32)
+        } else {
+            origin_date + Months::new(month_delta as u32)
+        };
+    }
```
this makes `2000-01-31T00:00:00` becomes `2000-01-29T00:00:00` in 
`❯ select date_bin('1 month', timestamp '2000-01-31T00:00:00', timestamp '2000-02-29T00:00:00');`

NGA-TRAN replied:
See https://github.com/apache/arrow-datafusion/pull/5982#issuecomment-1511640259

stuartcarnie reviewed (approved):
Looks great, @NGA-TRAN!

Not required for this PR, but perhaps something to consider in a followup is that I thought I would experiment with caching the `stride` function based on the `interval` input to the `date_bin` function:

https://github.com/NGA-TRAN/arrow-datafusion/blob/0dc43290387b7eadb24dcf7c9fcd3576ab3f2672/datafusion/physical-expr/src/datetime_expressions.rs#L511

I came up with the following change, that includes a benchmark of `date_bin`.

For the 1,000 element timestamp array the baseline (uncached):

```text
month-day-nano/date_bin/1000
                        time:   [3.7791 µs 3.7870 µs 3.7958 µs]
                        thrpt:  [263.45 Melem/s 264.06 Melem/s 264.62 Melem/s]
```

and after the change, the throughput, which is the number of timestamps elements mapped by the function per second, improved consistently by about 28%:

```text
month-day-nano/date_bin/1000
                        time:   [2.9499 µs 2.9529 µs 2.9563 µs]
                        thrpt:  [338.26 Melem/s 338.65 Melem/s 339.00 Melem/s]
                 change:
                        time:   [-22.338% -22.151% -21.964%] (p = 0.00 < 0.05)
                        thrpt:  [+28.147% +28.454% +28.763%]
                        Performance has improved.
```

```patch
Subject: [PATCH] feat: cache stride function
---
Index: datafusion/physical-expr/Cargo.toml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/datafusion/physical-expr/Cargo.toml b/datafusion/physical-expr/Cargo.toml
--- a/datafusion/physical-expr/Cargo.toml	(revision 0dc43290387b7eadb24dcf7c9fcd3576ab3f2672)
+++ b/datafusion/physical-expr/Cargo.toml	(date 1681677085430)
@@ -75,3 +75,7 @@
 [[bench]]
 harness = false
 name = "in_list"
+
+[[bench]]
+harness = false
+name = "date_bin"
Index: datafusion/physical-expr/benches/date_bin.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/datafusion/physical-expr/benches/date_bin.rs b/datafusion/physical-expr/benches/date_bin.rs
new file mode 100644
--- /dev/null	(date 1681679172908)
+++ b/datafusion/physical-expr/benches/date_bin.rs	(date 1681679172908)
@@ -0,0 +1,38 @@
+use arrow_array::TimestampNanosecondArray;
+use criterion::{
+    black_box, criterion_group, criterion_main, BenchmarkId, Criterion, Throughput,
+};
+use datafusion_common::ScalarValue;
+use datafusion_expr::ColumnarValue;
+use datafusion_physical_expr::datetime_expressions::date_bin;
+use std::sync::Arc;
+
+fn bench_mdn(c: &mut Criterion) {
+    let elem_10 =
+        TimestampNanosecondArray::from_iter_values((0..(10 as i64)).into_iter());
+    let elem_1000 =
+        TimestampNanosecondArray::from_iter_values((0..(1000 as i64)).into_iter());
+
+    let mut group = c.benchmark_group("month-day-nano");
+    for elements in [Arc::new(elem_10), Arc::new(elem_1000)].iter() {
+        group.throughput(Throughput::Elements(elements.len() as u64));
+
+        let args = vec![
+            ColumnarValue::Scalar(ScalarValue::new_interval_mdn(0, 1, 0)),
+            ColumnarValue::Array(elements.clone()),
+            ColumnarValue::Scalar(ScalarValue::TimestampNanosecond(Some(1), None)),
+        ];
+
+        group.bench_with_input(
+            BenchmarkId::new("date_bin", elements.len()),
+            &args,
+            |b, args| {
+                b.iter(|| date_bin(args));
+            },
+        );
+    }
+    group.finish();
+}
+
+criterion_group!(benches, bench_mdn);
+criterion_main!(benches);
Index: datafusion/physical-expr/src/datetime_expressions.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/datafusion/physical-expr/src/datetime_expressions.rs b/datafusion/physical-expr/src/datetime_expressions.rs
--- a/datafusion/physical-expr/src/datetime_expressions.rs	(revision 0dc43290387b7eadb24dcf7c9fcd3576ab3f2672)
+++ b/datafusion/physical-expr/src/datetime_expressions.rs	(date 1681679592855)
@@ -428,6 +428,13 @@
             Interval::Months(months) => date_bin_months_interval(*months, source, origin),
         }
     }
+
+    fn interval_fn(&self) -> (i64, fn(i64, i64, i64) -> i64) {
+        match self {
+            Interval::Nanoseconds(nanos) => (*nanos, date_bin_nanos_interval),
+            Interval::Months(months) => (*months, date_bin_months_interval),
+        }
+    }
 }
 
 // Supported intervals:
@@ -508,7 +515,8 @@
         )),
     };
 
-    let f = |x: Option<i64>| x.map(|x| stride.bin(x, origin));
+    let (stride, stride_fn) = stride.interval_fn();
+    let f = |x: Option<i64>| x.map(|x| stride_fn(stride, x, origin));
 
     Ok(match array {
         ColumnarValue::Scalar(ScalarValue::TimestampNanosecond(v, tz_opt)) => {
```

NGA-TRAN commented:
@waitingkuo 

You have great tests. I agree they look inconsistent. I think we may want to define the behavior here and see whether we want to keep it or not:

I think this one is easy to understand: the origin is Jan 31 so any other month will always be the last day of that month (28, 29, 30, 31 depending on  the month and year)

```SQL
❯ select date_bin('1 month', timestamp '2000-02-29T00:00:00', timestamp '2000-01-31T00:00:00');
+----------------------------------------------------------------------------------+
| datebin(Utf8("1 month"),Utf8("2000-02-29T00:00:00"),Utf8("2000-01-31T00:00:00")) |
+----------------------------------------------------------------------------------+
| 2000-02-29T00:00:00                                                              |
+----------------------------------------------------------------------------------+
1 row in set. Query took 0.003 seconds.
```

This is is tricky and we need to identify the behavior.  

```SQL
❯ select date_bin('1 month', timestamp '2000-01-31T00:00:00', timestamp '2000-02-29T00:00:00');
+----------------------------------------------------------------------------------+
| datebin(Utf8("1 month"),Utf8("2000-01-31T00:00:00"),Utf8("2000-02-29T00:00:00")) |
+----------------------------------------------------------------------------------+
| 2000-01-29T00:00:00                                                              |
+----------------------------------------------------------------------------------+
1 row in set. Query took 0.003 seconds.
```

For February, 29 (or 28) is the last day of the month, but for other month is it not the last day of the month. So chromo's [Months](https://pola-rs.github.io/polars/chrono/struct.Months.html) that we are using in this PR do not consider 29 as the last day of the month and just simple add/minus 30 days which I think makes sense.

```SQL
❯ select date_bin('1 month', timestamp '2000-01-31T00:00:00', timestamp '2000-03-29T00:00:00');
+----------------------------------------------------------------------------------+
| datebin(Utf8("1 month"),Utf8("2000-01-31T00:00:00"),Utf8("2000-03-29T00:00:00")) |
+----------------------------------------------------------------------------------+
| 2000-01-29T00:00:00                                                              |
+----------------------------------------------------------------------------------+
1 row in set. Query took 0.005 seconds.
```

So I think this is the behavior is acceptable. What do you, @alamb and @stuartcarnie think?

NGA-TRAN commented:
I am adding @stuartcarnie 's suggestion and benchmark into this PR

NGA-TRAN commented:
@alamb : I have added @stuartcarnie's the suggestion without the benchmark because I got Audit error if the benchmark is included. I am not sure what is the policy for adding benchmark

stuartcarnie commented:
That's great, thanks @NGA-TRAN! `date_bin` is used extensively in IOx, so downstream will definitely appreciate it!

NGA-TRAN commented:
I have added the tests with origin on 29th of Feb and Mar and interval is a month. I have added explanation for the tests on why the bins are always on the 29th

alamb reviewed (approved):
Thanks everyone for their comments and help here
